### ISSUE TITLE
Support easily parsable ranges

### ISSUE THREAD

**Issue by sholladay** [zaptst/zap#1]
Line and column ranges would be useful for linters and other tools that want to call attention to specific areas of code.

**Comment by jamiebuilds** [comment]
https://github.com/zaptst/zap/blob/master/README.md#source The `source` field has a line a column which could be expanded to be a range:


```
path
path:line
path:line-line
path:line:column
path:line:column-line:column
```

Many others have used this format before, it can even be expanded for allowing multiple source locations if needed.

Is this enough or was there something more you were looking for?

**Comment by sholladay** [comment]
Oops. I was clearly reading too fast, sorry! I thought `source` was only the filepath.

That said, ranges are very useful for linters. It allows me to report an issue with a particular function parameter, for example, and have the reporter highlight the whole parameter intuitively. If multiple ranges are supported, it allows me to highlight unbalanced parentheses and other issues that are not side-by-side in the code.

On a semi-related note, I think there should be a separate `position` field for this data, to reduce the need for string parsing. And if the column is missing, it means it applies to the whole line. In the example below, I will pretend we have that.

```js
{
    "position" : [
        "3",
        "6:9-8"
    ]
}
```

... would mean all of line 3, nothing from line 4 or 5, and then from line 6 column 9 through all of line 8, including the rest of line 6 and all of line 7.

I would consider breaking this down into objects to further reduce string parsing, but that can be a separate issue.

**Comment by jamiebuilds** [comment]
Okay, I want to know all the requirements first though, like can you have multiple source positions across files? For tools like type checkers you absolutely will all the time, probably linters too.

But I'm not sure if there's much of a reason for linters or type checkers to use a format like this. Do any linters use JUnit or TAP output? It'd be a confusing use at the very least

**Comment by sholladay** [comment]
> can you have multiple source positions across files

I can just emit multiple events, one per file, to represent that. i don't think the format needs anything special to say a single event applies to multiple files. A reporter can always map them to one event by comparing the `kind`, `id`, `message`, etc. of multiple events.

> Do any linters use JUnit or TAP output?

Absolutely! ESLint supports [JUnit](https://eslint.org/docs/user-guide/formatters/#junit) and [TAP](https://eslint.org/docs/user-guide/formatters/#tap) and it has all of the downsides you mention in the README. I don't find it confusing at all. I think linters desperately need better output formats like ZAP.

I find it annoying that when there are lint violations on a PR, I have to go to CI and read through a long list of relatively unstructured output in the CI terminal and then correlate that with the source code. It should behave more like a code review and just show me the errors inline on the PR. We're getting closer to that with the [GitHub Checks API](https://blog.github.com/2018-05-07-introducing-checks-api/) and I think ZAP could help by having first-class support for ranges and representing them in a machine-friendly manner.

Also, I would consider using this in [mop](https://github.com/sholladay/mop-cli), a linter of sorts that checks multiple projects rather than just one. I plan to eventually make a GUI for it and I think `mop` may need to go all-in on streams, because there can potentially be a ton of output.

**Comment by jamiebuilds** [comment]
> Absolutely! ESLint supports JUnit and TAP and it has all of the downsides you mention in the README. I don't find it confusing at all. I think linters desperately need better output formats like ZAP.

Okay interesting, I hadn't considered that. I've updated the README to include details about supporting linters/etc.

> I can just emit multiple events, one per file, to represent that.

I'm not sure that sending multiple events is the best solution here. If two events should be reported together, you would have to wait to report the first one, and the only way to fold them together right now would be to use `.message`.

I want to try and keep the ZAP events as flat as possible and not bloat them too much. And by making source locations their own format you can make them more condensed and defer the need to parse them if you don't care about them.

The grammar for could be:

```js
path:row:col-row:col,row,col-row:col;path:row:col-row:col,row,col-row:col
```

Which is very easy to parse, for example in JavaScript:

```js
function parseLoc(str) {
  let [row, col] = str.split(':');
  return {
    row: row ? parseInt(row) : null,
    col: col ? parseInt(col) : null,
  };
}

function parseRange(str) {
  let [start, end] = str.split('-');
  return {
    start: start ? parseLoc(start) : null,
    end: end ? parseLoc(end) : null,
  };
}

function parseFile(fileStr) {
  let [path, rangesStr] = fileStr.split(/:(.*)/);
  return {
    path: path,
    ranges: rangesStr ? rangesStr.split(',').map(parseRange) : [],
  };
}

function parseSource(source) {
  return source.split(';').map(parseFile);
}


let testCases = [
  'path/to/file.js',
  'path/to/file.js:',
  'path/to/file.js:42',
  'path/to/file.js:42:11',
  'path/to/file.js:-42',
  'path/to/file.js::42',
  'path/to/file.js:42:11-47',
  'path/to/file.js:42:11-47:15',
  'path/to/file.js:42:11-47:15,57',
  'path/to/file.js:42:11-47:15,57:4',
  'path/to/file.js:42:11-47:15,57:4-98:3',
  'path/to/file.js;path/to/file2.js',
  'path/to/file.js:42;path/to/file2.js',
  'path/to/file.js:42:11-47:15,57:4-98:3;path/to/file2.js',
  'path/to/file.js:42:11-47:15,57:4-98:3;path/to/file2.js:12',
  'path/to/file1.js:42:11-47:15,57:4-98:3;path/to/file2.js:12:6-15:4,20:8-24:10',
];

for (let testCase of testCases) {
  console.log(testCase, parseSource(testCase));
}
```

**Comment by sindresorhus** [comment]
I also think the source positions should be a separate field.

**Comment by jussi-kalliokoski** [comment]
One thing to consider is that not all source locations should be treated similarly in UIs, for example one use case is the unbalanced parens situation where the linter might know where you should put the parens and inform of the pair as context.

A similar scenario, but slightly different, is type checking where you can have a mismatch between definition and uses, and the editing may be required in the definition, in the use locations or both.

A third use case is error stack traces, where you’ll want to be able to go to any of the locations. One interesting thing about this scenario is also that the tool may have additional contextual information about the stack frames, such as variable values.

Considering all this, I think flat string source locations will not scale very well, and I suggest starting with just adopting LSP’s Location type, possibly with a wrapper object to account for extensions such as the stack frame context.

EDIT: also the type checking scenario possibly has different error messages per source location, so it makes sense to have a wrapper object.

**Comment by jamiebuilds** [comment]
I've opened up https://github.com/zaptst/zap/pull/6 to merge `"message"` and `"source"` with a `"content"` type which makes events able to describe many types of errors/warnings/etc. and makes source locations plain JSON values.

**Comment by jamiebuilds** [comment]
Resolved in #6
